processor/src/hasher/mod.rs

```diff
@@ -1,113 +1,335 @@
+
+        // determine values for the node index column for this permutation. if the first selector
+        // of init_selectors is not ZERO (i.e., we are processing the first leg of the Merkle
+        // path), the index for the first row is different from the index for the other rows;
+        // otherwise, indexes are the same.
+        let (init_index, rest_index) = if init_selectors[0] == Felt::ZERO {
+            (Felt::new(*index >> 1), Felt::new(*index >> 1))
+        } else {
+            (Felt::new(*index), Felt::new(*index >> 1))
+        };
+
+        // apply the permutation to the state and record its trace
+        self.trace.append_permutation(
+            &mut state,
+            init_selectors,
+            final_selectors,
+            init_index,
+            rest_index,
+        );
+
+        // remove the least significant bit from the index and return hash result
+        *index >>= 1;
+        [state[4], state[5], state[6], state[7]]
+    }
+}
+
+// MERKLE PATH CONTEXT
+// ================================================================================================
+
+/// Specifies the context of a Merkle path computation.
+enum MerklePathContext {
+    /// The computation is for verifying a Merkle path (MPVERIFY).
+    MpVerify,
+    /// The computation is for verifying a Merkle path to an old node during Merkle root update
+    /// procedure (MRUPDATE).
+    MrUpdateOld,
+    /// The computation is for verifying a Merkle path to a new node during Merkle root update
+    /// procedure (MRUPDATE).
+    MrUpdateNew,
+}
+
+impl MerklePathContext {
+    /// Returns selector values for this context.
+    pub fn main_selectors(&self) -> Selectors {
+        match self {
+            Self::MpVerify => MP_VERIFY,
+            Self::MrUpdateOld => MR_UPDATE_OLD,
+            Self::MrUpdateNew => MR_UPDATE_NEW,
         }
+    }
 
-        // TODO: return address of the hash table row
-        (Felt::ZERO, old_root, new_root)
+    /// Returns partial selector values for this context. Partial selector values are derived
+    /// from selector values by replacing the first selector with ZERO.
+    pub fn part_selectors(&self) -> Selectors {
+        let selectors = self.main_selectors();
+        [Felt::ZERO, selectors[1], selectors[2]]
     }
 }
 
 // HELPER FUNCTIONS
 // ================================================================================================
 
-fn build_merge_state(a: &Word, b: &Word) -> HasherState {
-    [
-        Felt::new(8),
-        Felt::ZERO,
-        Felt::ZERO,
-        Felt::ZERO,
-        a[0],
-        a[1],
-        a[2],
-        a[3],
-        b[0],
-        b[1],
-        b[2],
-        b[3],
-    ]
+/// Combines two words into a hasher state for Merkle path computation.
+///
+/// If index_bit = 0, the words are combined as a | b, if index_bit = 1, the words are combined
+/// as b | a, otherwise, the function panics.
```

Comment:
nit: maybe this comment could be clearer. To me, it reads as a bitwise ```OR```, but it seems to be about the order in which the elements are placed in the hasher state array

Reply:
Yep - re-worded.